### ISSUE TITLE
[ENH] Add truncate_date function

### ISSUE THREAD

**Issue by szuckerman** [pyjanitor-devs/pyjanitor#198]
I truncate dates all the time and it's really annoying. I want to add a `date_trunc` method.

**Comment by lphk92** [comment]
To be consistent with the other functions in this library, maybe it should be called `truncate_date`?

**Comment by ericmjl** [comment]
@szuckerman, do you have a simple example that illustrates what you mean by truncating a date?

**Comment by szuckerman** [comment]
First, the reason I wrote `date_trunc` before, rather than `truncate_date` is because of [the common SQL function by the same name](http://www.postgresqltutorial.com/postgresql-date_trunc/).

I like the new suggestion, but maybe include `date_trunc` as an alias as well?

An example why this is useful:

Say I have weblogs that have `user_id` and `view_date` as values. `view_date` may have precision the the nearest millisecond. If I want to find trends or make a graph over days/months/years, I truncate `view_date` to the nearest day, month, or year. After that new column contains all the truncated dates, I can `groupby` that column and `count` the number of `user_id`s to see how many users interacted by day/month/year.

Basically, it helps for date aggregations.

**Comment by lphk92** [comment]
So to make sure I'm understanding, an example implementation of what you're talking about would be something like

```
def truncate_date(df, column):
    df[column] = df[column].dt.date()
    return df```

**Comment by szuckerman** [comment]
Here's some example code for truncating by month. You'd obviously need to change it to match the `pyjanitor` methods:

```python
from datetime import date

def month_trunc(x):
    return date(x.year, x.month, 1)

dat['date_month'] = dat.date_value.apply(month_trunc)

dat_agg = dat.groupby('date_month')[['users']].sum().reset_index()

```

**Comment by lphk92** [comment]
I think in general we should avoid using `.apply()` where at all possible, since it is typically the slowest method for doing any transformation on a DataFrame. Do you think it's worth having separate functions for year, month, and day truncation?

**Comment by szuckerman** [comment]
I think it should mimic the `SQL` `date_trunc` I [referenced previously](http://www.postgresqltutorial.com/postgresql-date_trunc/). There are 13 truncations there, but I would guess that the most important are year, month, day, hour (and maybe quarter), and have those options be a parameter.

**Comment by szuckerman** [comment]
On second thought, maybe this function should be changed to something like `floor_date`. That way you can also have parameters for things like `first_day_of_week` in addition to `year`, `month`, etc.

The [date_util](https://dateutil.readthedocs.io/en/stable/) package might help with some of the work.
Similarly, here's [a stackoverflow question](https://stackoverflow.com/questions/42285130/how-floor-a-date-to-the-first-date-of-that-month) that shows some examples of how people have dealt with floor_dates in `pandas`.

**Comment by szuckerman** [comment]
I think this is still useful, but I forgot about the [`pd.Grouper`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Grouper.html) function that accomplishes this as well for a `DataFrame`. It accepts various frequencies for which to group by; anything mentioned on [this page](http://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html).

One thing to think about, though, is that using `pd.Grouper` is once one is performing an analysis and the data is already cleaned; the `truncate_date` function is to initially clean the data.

Thoughts?

**Comment by ericmjl** [comment]
This is interesting!

I wanted to leave this thought here, even though I am currently lacking the words for the context surrounding the thought: it'd possibly go well with `groupby_agg`, which was contributed at the sprints.
